File: 操作系统/2024-11-14/job.py
#-*- coding: utf-8 -*-
import threading
import time

# 初始化
empty_slots = threading.Semaphore(40)  # 果盘中空数量，最多40个
apple_in_basket = threading.Semaphore(0)  # 果盘中没有苹果
orange_in_basket = threading.Semaphore(0)  # 果盘中没有橘子
mutex = threading.Semaphore(1)  # 互斥

# 爸爸不停放苹果
def dad():
    while True:
        empty_slots.acquire()  # 确保果盘中
        mutex.acquire()  # 互斥
        print("爸爸放一个苹果")
        apple_in_basket.release()  # 放入
        mutex.release()  # 释放
        time.sleep(1)

# 妈妈不停放橘子
def mom():
    while True:
        empty_slots.acquire()
        mutex.acquire()  # 互斥
        print("妈妈放一个橘子")
        orange_in_basket.release()  # 放入
        mutex.release()  # 释放
        time.sleep(1)

# 儿子不停取橘子
def son():
    while True:
        orange_in_basket.acquire()
        mutex.acquire()
        print("儿子取一个橘子")
        empty_slots.release()
        mutex.release()
        time.sleep(1)

# 女儿的线程：不停取苹果
def daughter():
    while True:
        apple_in_basket.acquire()
        mutex.acquire()
        print("女儿一个苹果")
        empty_slots.release()
        mutex.release()
        time.sleep(1)

# 创建线程
dad_thread = threading.Thread(target=dad)
mom_thread = threading.Thread(target=mom)
son_thread = threading.Thread(target=son)
daughter_thread = threading.Thread(target=daughter)


dad_thread.start()
mom_thread.start()
son_thread.start()
daughter_thread.start()

dad_thread.join()
mom_thread.join()
son_thread.join()
daughter_thread.join()

File: 操作系统/2024-11-14/job.c
、#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>
#include <semaphore.h>
#include <unistd.h>

#define MAX_FRUITS 5  // 果盘容量

// 信号量
sem_t empty;  // 果盘空位
sem_t full;   // 果盘中水果数量
sem_t mutex;  // 互斥信号量

int fruit_basket[MAX_FRUITS];  
int fruit_count = 0;            // 当前数量

// 放水果函数
void put_fruit(int type) {
    fruit_basket[fruit_count] = type;
    fruit_count++;
    printf("放了一个水果, 当前水果数量：%d\n", fruit_count);
}

// 取水果函数
void take_fruit(int* fruit_count_local) {
    (*fruit_count_local)--;
    printf("取了一种水果, 当前水果数量：%d\n", *fruit_count_local);
}

// 爸爸放苹果
void* dad(void* arg) {
    while (1) {
        sem_wait(&empty);  // 等待空位
        sem_wait(&mutex);  // 临界区

        // 放苹果
        if (fruit_count < MAX_FRUITS) {
            put_fruit(1);  // 1表示苹果
        }

        sem_post(&mutex);  // 离开临界区
        sem_post(&full);   // 增加已放的数量

        sleep(1);  // 模拟放时间
    }
    return NULL;
}

// 妈妈放橘子
void* mom(void* arg) {
    while (1) {
        sem_wait(&empty);  
        sem_wait(&mutex);  


        if (fruit_count < MAX_FRUITS) {
            put_fruit(2); 
        }

        sem_post(&mutex);  
        sem_post(&full);  

        sleep(1); 
    }
    return NULL;
}

// 取橘子
void* son(void* arg) {
    while (1) {
        sem_wait(&full);  // 等待果盘有水果
        sem_wait(&mutex);  // 进临界区

        // 取橘子
        if (fruit_count > 0) {
            take_fruit(&fruit_count);
        }

        sem_post(&mutex);  // 离开临界区
        sem_post(&empty);  // 增加空位数

        sleep(1);  // 模拟取水果的时间
    }
    return NULL;
}


void* daughter(void* arg) {
    while (1) {
        sem_wait(&full);  
        sem_wait(&mutex); 


        if (fruit_count > 0) {
            take_fruit(&fruit_count);
        }

        sem_post(&mutex); 
        sem_post(&empty);  

        sleep(1);  
    }
    return NULL;
}

int main() {
    pthread_t dad_thread, mom_thread, son_thread, daughter_thread;

    // 初始化信号量
    if (sem_init(&empty, 0, MAX_FRUITS) != 0 ||
        sem_init(&full, 0, 0) != 0 ||
        sem_init(&mutex, 0, 1) != 0) {
        perror("信号量初始化失败");
        exit(EXIT_FAILURE);
    }

    // 创建线程
    if (pthread_create(&dad_thread, NULL, dad, NULL) ||
        pthread_create(&mom_thread, NULL, mom, NULL) ||
        pthread_create(&son_thread, NULL, son, NULL) ||
        pthread_create(&daughter_thread, NULL, daughter, NULL)) {
        perror("线程创建失败");
        exit(EXIT_FAILURE);
    }

    // 等待线程结束
    pthread_join(dad_thread, NULL);
    pthread_join(mom_thread, NULL);
    pthread_join(son_thread, NULL);
    pthread_join(daughter_thread, NULL);

    // 销毁信号量
    sem_destroy(&empty);
    sem_destroy(&full);
    sem_destroy(&mutex);

    return 0;
}
